RHOAIENG-18979: chore(test/containers): check workbench startup logs

jstourac opened the pull request:
This adds a new check for the content of the workbench startup logs searching for some keywords that may indicate some problem or issue.

https://issues.redhat.com/browse/RHOAIENG-18979

<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
```
DOCKER_HOST=unix:///run/user/$UID/podman/podman.sock poetry run pytest tests/containers --image quay.io/modh/odh-generic-data-science-notebook@sha256:0efbb3ad6f8f342360cf1f002d40716a39d4c58f69163e053d5bd19b4fe732d4 -k test_image_entrypoint_starts

DOCKER_HOST=unix:///run/user/$UID/podman/podman.sock poetry run pytest tests/containers --image quay.io/opendatahub/workbench-images@sha256:27d87f50f7aad6185344368b1b00c446773cff12bbea2f87e8d0c4ece2b13d87 -k test_image_entrypoint_starts
```

## Merge criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

jiridanek commented on tests/containers/workbenches/workbench_image_test.py:
```diff
@@ -31,21 +32,48 @@ class TestWorkbenchImage:
+                        for keyword in blocked_keywords:
+                            if keyword in line:
+                                if not any(allowed in line for allowed in allowed_messages):
+                                    logging.debug(f"Unexpected keyword '{keyword}' in the following message: '{line}'")
```
logging.info? or maybe logging.warning?

jstourac replied:
you are right, I updated it to `error` actually

jiridanek commented on tests/containers/workbenches/workbench_image_test.py:
```diff
@@ -31,21 +32,48 @@ class TestWorkbenchImage:
+
+                # let's check that logs don't contain any error or unexpected warning message
+                with subtests.test("Checking the log in the workbench..."):
+                    failed_lines = 0
```
I have three ideas to avoid counting lines ourselves

* when there is a failure, do a `with subtests.test(): assert`, that is, create a subtest for every failed test
* there is https://github.com/okken/pytest-check
* and there is https://github.com/astraw38/pytest-assume

don't really have a preference, this counter is not exactly a bad solution either

jstourac replied:
Okay, I kept this as one test since we probably don't want different number of subtests depending on how many log messages are there :thinking: I chose to alternate this part of code based on your [other suggestion](https://github.com/opendatahub-io/notebooks/pull/886#discussion_r1942769221).

jiridanek commented on tests/containers/workbenches/workbench_image_test.py:
```diff
@@ -31,21 +32,48 @@ class TestWorkbenchImage:
+                with subtests.test("Checking the log in the workbench..."):
+                    failed_lines = 0
+                    for line in full_logs.splitlines():
+                        for keyword in blocked_keywords:
+                            if keyword in line:
+                                if not any(allowed in line for allowed in allowed_messages):
+                                    logging.debug(f"Unexpected keyword '{keyword}' in the following message: '{line}'")
+                                    failed_lines += 1
+                                    # Let's go on another line now as we don't care if multiple blocked keywords are present in this line
+                                    break
```
just this small idea that saves one level of indentation and comment is no longer necessary imo

```suggestion
                        if any(keyword in line for keyword in blocked_keywords):
                            if not any(allowed in line for allowed in allowed_messages):
                                logging.debug(f"Unexpected keyword '{keyword}' in the following message: '{line}'")
                                failed_lines += 1
```

jstourac replied:
This has a disadvantage as I am not able to print the keyword that caused this to fail... but yeah, I'll probably amend that this way.

jstourac replied:
that would be too verbose... let's not complicate it for now, we can always improve

jiridanek replied:
```
[Checking the log in the workbench...] SUBFAIL tests/containers/workbenches/workbench_image_test.py::TestWorkbenchImage::test_image_entrypoint_starts[ghcr.io/opendatahub-io/notebooks/workbench-images:jupyter-trustyai-ubi9-python-3.11-889_merge_ad1bfa2899f33cd9962985d35165c6b6b3b5caf3-sysctls1] - Failed: Log message(s) (3) that violate our checks occurred during the workbench startup:
[W 2025-02-06 08:26:45.534 LabApp] Failed to instantiate the extension manager pypi. Falling back to read-only manager.
    Traceback (most recent call last):
    TypeError: AsyncClient.__init__() got an unexpected keyword argument 'proxies'
```

This is _good_ output, I like it. From https://github.com/opendatahub-io/notebooks/actions/runs/13174582770/job/36771027197?pr=889#step:25:151

jiridanek commented on tests/containers/workbenches/workbench_image_test.py:
```diff
@@ -31,21 +32,48 @@ class TestWorkbenchImage:
+    def test_image_entrypoint_starts(self, subtests: pytest_subtests.SubTests, image: str, sysctls) -> None:
         skip_if_not_workbench_image(image)
 
+        # Here is a list of blocked keywords we don't want to see in the log messages during the container/workbench
```
maybe put all this into a helper function (logs and subtests object can be a parameter to it) and also call it from the ipv6 only test below?

jiridanek commented on tests/containers/workbenches/workbench_image_test.py:
```diff
@@ -31,21 +32,48 @@ class TestWorkbenchImage:
+
+                # let's check that logs don't contain any error or unexpected warning message
+                with subtests.test("Checking the log in the workbench..."):
+                    failed_lines = 0
```
on second thought, maybe I'd make this

```suggestion
                    failed_lines = []
```

and .append to it the failed lines

why should people hunt for them in the log when they can be kept here and produced in the final failure message

jstourac replied:
yup, I like it - updated

jiridanek commented on tests/containers/workbenches/workbench_image_test.py:
```diff
@@ -31,21 +32,51 @@ class TestWorkbenchImage:
+                    container._connect()
             finally:
                 # try to grab logs regardless of whether container started or not
+                # TODO - do we want to enforce couple of seconds wait to be sure nothing extra suspicious is logged for the workbench???
```
either that, or possibly we may even want to try to use jupyterlab a bit

every playwright test that we're going to eventually create could be checking logs like this, even, maybe

future work

andyatmiami replied:
minor nit - but i thought we were told we needed to write cypress tests (?)

jiridanek replied:
jupyterlab and codeserver upstreams use playwright

elyra and odh dashboard use cypress

dunno about rstudio

most importantly we need to use what's already present

jstourac replied:
okay, so I put a 3 second wait there for now... we may alternate this in the future or remove eventually

jstourac commented:
Thank you for your preliminary review, guys. I addressed all comments hopefully and marked this as ready for review now.

Also, I'll wait with merging this until the https://issues.redhat.com/browse/RHOAIENG-18930 is addressed also, otherwise we would have failing tests in our CI in the meantime.

jiridanek commented:
/lgtm

jstourac commented:
Okay, so the fix for the Jupyter is prepared here #889. I would go with the following steps:
1. merge this thing
2. then rebase the #889 and follow there

atheo89 commented:
/lgtm

jstourac commented:
Thank you for your reviews and comments, guys!

/approve

jstourac commented:
/override ci/prow/images
/override "build (rocm-jupyter-tensorflow-ubi9-python-3.11) / build (ubuntu-22.04)"
/override "build (rocm-jupyter-pytorch-ubi9-python-3.11) / build (ubuntu-22.04)"
/override "build (jupyter-pytorch-ubi9-python-3.11) / build (ubuntu-22.04)"
/override "build (jupyter-trustyai-ubi9-python-3.11) / build (ubuntu-22.04)"
/override "build (cuda-jupyter-tensorflow-ubi9-python-3.11) / build (ubuntu-22.04)"
